### ISSUE TITLE
dataset

### ISSUE THREAD

**Issue by leosegre** [cchen-cc/SIFA#15]
Hi,

We would like to use your dataset, but we have some issues using it.
We are trying to convert the slices into numpy arrays, but there are mismatches between the dataset to the discription in the loader.py file.
The dataset have more fields like mask and bias.
The dataset also have more dimensions - the lables are 256X256X3 instead of 256X256X1, the data array is 256x256x3**x4**.

Can you please explain us how to read the dataset properly?

I attached our attempt to read it with the loader.py parameters.
![data read](https://user-images.githubusercontent.com/58905206/94663514-6b559e80-0312-11eb-9b86-9ffc27ee50e4.jpeg)

Thanks,
Leo

**Comment by cchen-cc** [comment]
Hi Leo,

The dataset should be read correctly by data_loader.py. 
I don't quite understand how you get label with size 256*256*3 and data array with size 256*256*3*4. If you used data_loader.py to read the data, you should get label with size 256*256*1 and data with size 256*256*3 as specified in the script. 
The figure you attached looks weird, and I cannot see any useful information.

Best,
Cheng

**Comment by leosegre** [comment]
Thank you for your answer,
We are using pytorch so we can't use data_loader.py directly.
When we try to load the data into dictionary we get parameters as in the pictures.
As you can see there are more parameters than in the loader.py.
Also, the data is 786,432 length.

![IMG-20200927-WA0004](https://user-images.githubusercontent.com/58905206/95105377-c1c24300-073f-11eb-95ba-a753996c1013.jpg)

**Comment by cchen-cc** [comment]
What function did you use to read the tfrecord data then?

**Comment by leosegre** [comment]
We used TFRecordDataset - [](https://github.com/vahidk/tfrecord)

**Comment by cchen-cc** [comment]
I think the problem is caused by the data type. 
In your screenshot, it shows that the dtype of the data_vol is torch.uint8, but actually the data type used to save data_vol is float32. That's why you have data with size 256x256x3x4. If you change the data type to float32, the size becomes 256x256x3. The dtype of the label_vol should also be float32.

**Comment by leosegre** [comment]
Thank you for your answer, we used your advise and we managed to visualize a slice.
We normalized the data to [0,1] by the equation (x-x_min)/(x_max-x_min) or (x-(-2.8))/(3.2-(-2.8)), but the picture has some color artifacts.
Also, it is not aligned to the 256X256 frame. we tried to concatenate all slices into a 3D image but it is not working properly.

Can you please share how to visualize the data and how to build a 3D image from the slices?
It seems they are not ordered and there are some pictures we don't recognize as cardiac.

Thanks,
Leo

after normalizing using max and min
![image](https://user-images.githubusercontent.com/58905206/95675006-b4cca600-0bbc-11eb-984e-3fd50d4464f6.png)

after normalizing using -2.8 and 3.2
![image](https://user-images.githubusercontent.com/58905206/95675025-e3e31780-0bbc-11eb-8fe3-e23d538cc504.png)

non-cardiac slice
![image](https://user-images.githubusercontent.com/58905206/95675091-36bccf00-0bbd-11eb-8391-a995518f01c4.png)

**Comment by cchen-cc** [comment]
To visualize a slice, there is no need to normalize it, you can directly use the raw data. Gray scale color map should be used as as it's a gray-scale image. 
How did you concatenate the slices?
The slices in the training set were generated from 16 scans with data augmentation. Directly concatenating them would not give one reasonable 3D scan.

**Comment by leosegre** [comment]
Thank you,
If it is a gray-scale image why is it 256×256×3 and not 256×256×1?
To visualize it as float32 the values should be [0,1] but it isn't do we normalized it.

We directly concatenated them and as you said it is not a reasonable 3D scan. So how can we get a reasonable 3D scan?
We want to use your 3D scans after you applied augmentetions on it

**Comment by leosegre** [comment]
Hi,

Can you please share your preprocess pipeline?
We are working directly on 3D scans.

Thanks,
Leo

**Comment by cchen-cc** [comment]
The data are saved with the size 256x256x3 as we want to use three consecutive slices to leverage more spatial information in volumetric data in our other two works (https://github.com/carrenD/Medical-Cross-Modality-Domain-Adaptation), but in SIFA, we didn't use three consecutive slices.

To visualize, you can manually normalize the float32 data to uint8 data with range [0 255], but you don't have to do so as the image visualization function in python would automatically do that if you chose gray scale colormap.

We didn't release the augmented 3D scans, you may want to ask the contact persons of (https://github.com/carrenD/Medical-Cross-Modality-Domain-Adaptation) about the 3D data.
